### PR TITLE
Add mysql_url_parameters property

### PR DESCRIPTION
**Pull request by alwintje**

### CODE REVIEW

File: src/main/java/controllers/ImportController.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.web.bind.annotation.*;
+import services.ProjectService;
+
+@RestController
+@RequestMapping("/import")
+public class ImportController {
+
+    private final ProjectService projectService;
+
+    public ImportController(ProjectService projectService) {
```
Comment (Elvinasl) [inline src/main/java/controllers/ImportController.java]: @Autowire missing

File: src/main/java/controllers/ImportController.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        this.projectService = projectService;
+    }
+
+    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
+    @ResponseStatus(HttpStatus.CREATED)
+    @Transactional
+    public Dataset importJson(@RequestHeader("api-key") String apiKey, @RequestBody Dataset data) {
```
Comment (Elvinasl) [inline src/main/java/controllers/ImportController.java]: Are you sure we want to send api-key as a header? I would keep it in a body or in request url.

Comment (Elvinasl) [inline src/main/java/controllers/ImportController.java]: I would say we should return ResponseEntity with proper message.

File: src/main/java/controllers/adapters/JsonAdapter.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.util.List;
+
+public class JsonAdapter implements DatasetAdapter {
+
+
+    @Override
+    public List<ColumnName> convert(String data) {
+        return null;
+    }
```
Comment (Elvinasl) [inline src/main/java/controllers/adapters/JsonAdapter.java]: I am not sure why these adapters are here? Whats the point of them?
And why are they in controllers package?

File: src/main/java/models/Behavior.java [hunk 4]
```diff
@@ -21,11 +22,14 @@
...
-    @ManyToOne(optional = false)
+    @ManyToOne(optional = false, fetch = FetchType.LAZY)
+    @JsonIgnore
     @NotNull
     private Row row;
 
-    @ManyToOne(optional = false)
+    @ManyToOne(optional = false, fetch = FetchType.LAZY)
+    @JsonIgnore
     @NotNull
```
Comment (Elvinasl) [inline src/main/java/models/Behavior.java]: FetchType.Lazy is set by default. 

Please don't use @JsonIgnore. I believe you should use DTOs if you want to return some data..

File: src/main/java/models/Cell.java [hunk 5]
```diff
@@ -24,16 +26,25 @@
...
 
-    @ManyToOne(optional = false)
+    @ManyToOne(optional = false, fetch = FetchType.LAZY)
+    @JsonIgnore
     @NotNull
     private ColumnName columnName;
 
-    @ManyToMany
+    @ManyToMany(fetch = FetchType.LAZY)
+    @JsonIgnore
```
Comment (Elvinasl) [inline src/main/java/models/Cell.java]: Same as in Behavior object. Don't use @JsonIgnore.

File: src/main/java/models/Cell.java [hunk 6]
```diff
@@ -24,16 +26,25 @@
...
-    @ManyToOne(optional = false)
-    @NotNull
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JsonIgnore
     private UserPreference userPreference;
+
+    public void addRow(Row row) {
+        if (rows == null) {
+            rows = new ArrayList<>();
+        }
```
Comment (Elvinasl) [inline src/main/java/models/Cell.java]: We can just initialise ArrayList normally. 
private List<Row> rows = new ArrayList<>();

This looks a bit over engineering. :)

File: src/main/java/models/Client.java [hunk 7]
```diff
@@ -30,6 +30,8 @@
     @Column(nullable = false)
     private String password;
 
-    @OneToMany(mappedBy="client")
+    @OneToMany(mappedBy = "client", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
+    @JsonIgnore
```
Comment (Elvinasl) [inline src/main/java/models/Client.java]: @JsonIgnore remove it pls.

File: src/main/java/models/ColumnName.java [hunk 8]
```diff
@@ -27,11 +28,20 @@
...
 
-    @OneToMany(mappedBy = "columnName")
+    @OneToMany(mappedBy = "columnName", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
+    @JsonIgnore
     private List<Cell> cells;
+
+    public void add(Cell cell) {
+        if (cells == null) {
+            cells = new ArrayList<>();
+        }
```
Comment (Elvinasl) [inline src/main/java/models/ColumnName.java]: Same code smell as before.

File: src/main/java/models/Dataset.java [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+                .flatMap(List::stream)
+                .collect(Collectors.toMap(ColumnName::getName, cn -> cn, (ColumnName x, ColumnName y) -> y));
+
+        List<ColumnName> columns = new ArrayList<>(headers.values());
+        Map<String, Cell> cells = new HashMap<>();
+
+        project.setColumnNames(columns);
+        columns.forEach((column) -> {
+            column.setProject(project);
+        });
```
Comment (Elvinasl) [inline src/main/java/models/Dataset.java]: You can replace all this with Java 8 syntax - streams.

File: src/main/java/models/Dataset.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+                        col.add(cell);
+                    }
+                    cell.addRow(dbRow);
+                    dbRow.addCell(cell);
+                });
+                dbRows.add(dbRow);
+            });
+
+            project.setRows(dbRows);
+        }
```
Comment (Elvinasl) [inline src/main/java/models/Dataset.java]: Add some comments here. It is really hard to follow..

File: src/main/java/repositories/DatabaseRepository.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    public List<T> list() {
+        return em.createQuery("select o from " + type.getName() + " o").getResultList();
+    }
+
+    @Transactional
+    public void remove(int id) {
+        T o = this.getById(id);
+        if (null != o) {
+            em.remove(o);
+        }
```
Comment (Elvinasl) [inline src/main/java/repositories/DatabaseRepository.java]: You could throw exception if object is not found.

File: src/main/java/services/BehaviorService.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
+package services;
+
+import models.Behavior;
+import org.springframework.stereotype.Service;
+import repositories.BehaviorRepository;
+
+import java.util.List;
+
+@Service
+public class BehaviorService implements DatabaseServiceInterface<Behavior> {
```
Comment (Elvinasl) [inline src/main/java/services/BehaviorService.java]: Interface seems to be completely unnecessary here.

File: src/main/java/services/CellService.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
+package services;
+
+import models.Cell;
+import org.springframework.stereotype.Service;
+import repositories.CellRepository;
+
+import java.util.List;
+
+@Service
+public class CellService implements DatabaseServiceInterface<Cell> {
```
Comment (Elvinasl) [inline src/main/java/services/CellService.java]: Interface seems to be unnecessary here.

File: src/main/java/services/CellService.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final CellRepository cellRepository;
+
+    public CellService(CellRepository cellRepository) {
```
Comment (Elvinasl) [inline src/main/java/services/CellService.java]: @Autowired

File: src/main/java/services/ColumnNameService.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import repositories.ColumnNameRepository;
+
+import java.util.List;
+
+@Service
+public class ColumnNameService implements DatabaseServiceInterface<ColumnName> {
+
+    private final ColumnNameRepository columnNameRepository;
+
+    public ColumnNameService(ColumnNameRepository columnNameRepository) {
```
Comment (Elvinasl) [inline src/main/java/services/ColumnNameService.java]: @Autowired, interface not needed

File: src/main/java/controllers/ImportController.java (hunk 2 above)
Comment (tchristerus) [reply src/main/java/controllers/ImportController.java]: We'll ask the lecturer tomorrow about the best case in this situation. And how it fits in with the security part.
